src/translate.py

```diff
@@ -3808,16 +3814,19 @@ def process_instr(instr: Instruction) -> None:
                 func_args.append(regs[register])
 
             # Add the arguments after a3.
-            # TODO: limit this and unify types based on abi.arg_slots
-            subroutine_args.sort(key=lambda a: a[1].value)
-            for arg in subroutine_args:
-                func_args.append(arg[0])
+            # TODO: limit this based on abi.arg_slots. If the function type is known
+            # and not variadic, this list should be empty.
+            for _, arg in sorted(subroutine_args.items()):
+                func_args.append(arg)
 
             if not fn_sig.params_known:
                 while len(func_args) > len(fn_sig.params):
                     fn_sig.params.append(FunctionParam())
-            for i, (arg_expr, param) in enumerate(zip(func_args, fn_sig.params)):
-                func_args[i] = as_type(arg_expr, param.type, True)
+                # We can only perform this naive zip when `params_known` is False, and
+                # we're assuming that each parameter is "basic" (<=4 bytes, no return struct, etc.)
+                # Otherwise, we would need to duplicate the logic from `function_abi` here.
```

Comment:
so basically, when params_known = False, a FunctionSignature doesn't represent a real function signature but rather the ABI representation of one. (maybe an observation worth noting in a comment somewhere?)

Reply:
The way I was picturing it, is when `params_known = False`, we assume that the real function signature resembles the "naive" ABI representation. It's hard to tell if the first argument is a return struct pointer, or if set of 3 `f32`s on the stack are separate arguments or a single `Vec3f`.

I'll try to clarify this, though.